### ISSUE TITLE
calculate partial derivatives instead of total derivative like tf.gradients with parameter stop_gradients 

### ISSUE THREAD

**Issue by YunYunY** [pytorch/pytorch#46209]
## 🚀 Feature
calculate partial derivatives instead of total derivative like tf.gradients with parameter stop_gradients https://www.tensorflow.org/api_docs/python/tf/gradients

## Motivation
It is a useful feature to investigate the vanishing/exploding gradient problem like RNN. Usually, we want to check the last hidden layer partial gradient w.r.t the first hidden in RNN.

## Pitch
Same behavior like  tf.gradients with parameter stop_gradients.
```
a = torch.tensor(1., requires_grad=True)

b = 2 * a

c = a + b

print(torch.autograd.grad(c, a, retain_graph=True))
```
Output `(tensor(3.),)`, Expected output: 1 for partial derivative.
I know we can use `b = 2*(a.detach())`. But what if `d=2*c`. In that case, we want to keep calculating the total derivative dd/da. Expect to get `dd/da=6` without `a.detach` instead of 2.

Background: I am trying to calculate the last hidden layer partial gradient w.r.t the first hidden in RNN. After that, I still need to calculate loss.backward() where I don’t want to detach the hidden layer.

## Alternatives

Maybe using hook? But hook may not work in this case since hook only keeps the gradient of the input of nn.Module object w.r.t to the loss right? 

Thank you.

cc @ezyang @albanD @zou3519 @gqchen @pearu @nikitaved

**Comment by ezyang** [comment]
For clarity, OP is NOT referring to `tf.stop_gradient`, which is indeed the same thing as detach. If I understand correctly, you could work around this problem by manually detaching all occurrences of `a`; and this is annoying to do because you might have a lot of occurrences of a (and also you might want to still form the graph for a because you might want to compute a different partial derivative).

I think hooks will work for what you need, though still not quite as conveniently as the `stop_gradients` parameter. The hook can prevent gradients from being propagated which is what you want here.

**Comment by albanD** [comment]
As Ed mentionned just above. You can add a hook that looks like the following for every Tensor you want to "stop gradients" for: `t.register_hook(lambda grad: torch.zeros_like(grad))`.
You can then remove the hook when you want to do a full backward.

Does that match what you're trying to do?

**Comment by albanD** [comment]
I think we should move debugging discussions that are unrelated to the feature request to the forum: https://discuss.pytorch.org/
You can open a topic there with code sample and what you expect to see and ping me.

For the particular issue, didn't you get a warning when you tried to access the `.grad`?
From looking at the code, it looks like `states[0][1]` is the `h` that was given as input which does not require gradients. So it is expected that the grad is None.
